src/architecture/webgpu.md

```diff
@@ -0,0 +1,74 @@
+
+- DOM implementation
+- WGPU implementation
+
+```mermaid
+flowchart LR
+    subgraph Content Process
+        Script[Script Thread]
+    end
+
+    subgraph Main Process
+        WGPU[WGPU thread]
+        poller[WGPU poller]
+        WebRender
+        Constellation
+    end
+
+    %% init
+    Script--RequestAdapter-->Constellation
+    Constellation--RequestAdapter/WebRender-->WGPU
+    Script--WebGPURequest-->WGPU
+    WGPU--Adapter/WebGPU chan-->Script
+
+    %% normal operation
+    WGPU--token(), wake()-->poller
+    WGPU--Create/Update Image-->WebRender
+    WGPU--WebGPUResponse/WebGPUMsg-->Script
+```
+
+### DOM implementation
+
+DOM implementation lives in [`components/script/dom/webgpu`](https://github.com/servo/servo/tree/main/components/script/dom/webgpu) and implements JS interfaces defined in [WebGPU IDL](https://github.com/servo/servo/blob/main/components/script_bindings/webidls/WebGPU.webidl) that are exposed in browser's JS.
+
+Here we only implemented logic that is [described in spec under content timeline](https://www.w3.org/TR/webgpu/#content-timeline), which is mostly just converting JS to wgpu-types descriptors that are sent to WGPU thread via IPC messages that are defined in <https://github.com/servo/servo/blob/main/components/webgpu/ipc_messages/recv.rs>
+WebGPU is async by design, so there is no need to wait for WGPU thread operation to finish.
+This is done by storing ids in DOM WebGPU objects that links to WGPU objects that live in WGPU thread (are provided by wgpu-core). More information about this design is available in wgpu repo: <https://github.com/gfx-rs/wgpu/blob/2764e7a39920e23928d300e8856a672f1952da63/wgpu-core/src/hub.rs#L30>.
+
+### WGPU implementation
+
+Actual processing is done in separate thread (and separate process - Main Process): [`WGPU`](https://github.com/servo/servo/blob/main/components/webgpu/wgpu_thread.rs) and [`WGPU poller`](https://github.com/servo/servo/blob/main/components/webgpu/poll_thread.rs) thread that are started lazily on first request of adapter.
+
+WGPU thread handles incoming IPC messages from script and dispatches appropriate wgpu-core functions, that implement steps defined in [device timeline](https://www.w3.org/TR/webgpu/#device-timeline).
+Some calls also wake WGPU poller thread, which effectively runs [queue timeline](https://www.w3.org/TR/webgpu/#queue-timeline) by calling `poll_all_devices` from wgpu-core.
+
+## Updating WebGPU CTS expectations
+
+WebGPU CTS expectations are enormous, so instead of `mach update-wpt` we use [`moz-webgpu-cts`](https://github.com/erichdongubler-mozilla/moz-webgpu-cts) to update expectations (actually we use forked version <https://github.com/sagudev/moz-webgpu-cts/tree/servo>).
+
+For updating expectations, you need to obtain wptreport of test, either by triggering full CTS run in CI via `mach try webgpu` or by running specific tests locally `mach test-wpt -r --log-wptreport r.json {{tests}}`.
```

Comment:
```suggestion
To update expectations, you need to obtain the `wptreport` log for the run, either by triggering a full CTS run in CI (`mach try webgpu`) or by running specific tests locally:

---```sh (remove the dashes)
$ mach test-wpt -r --log-wptreport report.json [tests ...]
---``` (remove the dashes)
```